> > I'm surprised that CMake is trying to use targets defined in the same build in `try_compile`. What would happen if this a `LIBRARY` target like `unwind_shared` or `unwind_static`? That target hasn't been built yet (it's built by the build we're configuring). Is this a CMake issue or is there a well defined behavior for this case?
> 
> I tested replacing this with a check for `unwind_shared`, and that proceeded just fine (and it tested linking right away even if the target `unwind_shared` wasn't found).
> 
> Thus, if the `unwind` target wouldn't be of the type `UTILITY`, this might work - but would that cause other cmake confusion at some point? If the target/library entity `unwind` both means a preexisting library (link by passing `-lunwind`) and a library that gets built during the build (where it instead adds dependencies on it, builds that before using it, and links against the newly generated file instead of necessarily using literally `-lunwind`).

Thanks, I tried it locally as well to better understand the issue.

With the following:
```
cmake_minimum_required(VERSION 3.13.4)
project(Test LANGUAGES CXX)
add_library(unwind_static STATIC unwind.cpp)
add_custom_target(unwind DEPENDS unwind_static)
include(CheckLibraryExists)
check_library_exists(unwind _Unwind_RaiseException "" COMPILER_RT_HAS_LIBUNWIND)
```
I get the same error:
```
CMake Error at cmake/linux-x64/share/cmake-3.25/Modules/CheckLibraryExists.cmake:71 (try_compile):
  Only libraries may be used as try_compile or try_run IMPORTED
  LINK_LIBRARIES.  Got unwind of type UTILITY.
Call Stack (most recent call first):
  CMakeLists.txt:9 (check_library_exists)
```
However, with the following:
```
cmake_minimum_required(VERSION 3.13.4)
project(Test LANGUAGES CXX)
add_library(unwind STATIC unwind.cpp)
include(CheckLibraryExists)
check_library_exists(unwind _Unwind_RaiseException "" COMPILER_RT_HAS_LIBUNWIND)
```
CMake fails to find `libunwind`:
```
/usr/bin/ld: cannot find -lunwind: No such file or directory
```

To me this really looks like a bug in CMake since this behavior is inconsistent. Specifically, I don't think CMake should consider the previously defined targets when performing checks. Perhaps we should file an issue against CMake and include a link in the comment?

--- a/compiler-rt/cmake/config-ix.cmake
+++ b/compiler-rt/cmake/config-ix.cmake
@@ -63,6 +63,16 @@ if (C_SUPPORTS_NODEFAULTLIBS_FLAG)
                         moldname mingwex msvcrt)
     list(APPEND CMAKE_REQUIRED_LIBRARIES ${MINGW_LIBRARIES})
   endif()
+  if (NOT TARGET unwind)
+    # Don't check for a library named unwind, if there's a target with that name within
+    # the same build.
+    check_library_exists(unwind _Unwind_GetRegionStart "" COMPILER_RT_HAS_LIBUNWIND)
+    if (COMPILER_RT_HAS_LIBUNWIND)
+      # If we're omitting default libraries, we might need to manually link in libunwind.
+      # This can affect whether we detect a statically linked libc++ correctly.
+      list(APPEND CMAKE_REQUIRED_LIBRARIES unwind)
+    endif()
+  endif()
 endif ()
 
 # CodeGen options.

